### PR TITLE
Upgrade to Hono 1.2.1

### PR DESCRIPTION
**Pull request by dejanb**
### Type of change

- Refactoring

### Description

Upgrade to Hono 1.2.1

### CODE REVIEW

**Comment by dejanb** [comment]
@enmasse-ci run tests profile=iot

**Comment by dejanb** [comment]
@enmasse-ci run tests profile=iot

**Comment by dejanb** [comment]
@enmasse-ci run tests profile=iot

**Comment by dejanb** [comment]
@enmasse-ci run tests profile=iot

**Comment by dejanb** [comment]
@enmasse-ci run tests profile=iot

**Comment by dejanb** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.shared.control.CommandAndControlTest

**Comment by dejanb** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.shared.control.CommandAndControlTest

**Comment by dejanb** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.shared.control.CommandAndControlTest

**Comment by dejanb** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.shared.control.CommandAndControlTest

**Comment by dejanb** [comment]
@enmasse-ci run tests profile=iot

**Comment by dejanb** [comment]
This PR contains first work toward Hono 1.2 upgrade:

- Services have been refactored to use upstream base classes
- New interior router has been introduced and c&c has been fixed to use new addresses
- Temporary use mapped device connection service


Still left to do in subsequent PRs:

- Remove event bus adapters
- Properly instantiate and secure interior qdr with operator
- Rethink how to proceed with device connection service

File: iot/iot-device-registry-base/pom.xml [hunk 1]
```diff
@@ -63,6 +63,16 @@
...
       <artifactId>hamcrest</artifactId>
     </dependency>
+      <dependency>
+          <groupId>org.eclipse.hono</groupId>
+          <artifactId>hono-service-device-registry-base</artifactId>
+      </dependency>
+      <dependency>
+          <groupId>io.vertx</groupId>
+          <artifactId>vertx-junit5</artifactId>
+          <version>3.8.5</version>
```
Comment (ctron) [inline iot/iot-device-registry-base/pom.xml:73]: This should come from a parent `dependencyManagement` section.

File: iot/iot-device-registry-base/src/main/java/io/enmasse/iot/registry/tenant/KubernetesTenantInformationService.java [hunk 2]
```diff
@@ -2,56 +2,93 @@
...
-            return notFound(notFoundStatusCode, "Tenant does not exist");
+            return notFound(notFoundStatusCode, "Tenant '" + tenantName + "' does not exist");
         }
 
         final IoTProject project = projectResult.orElseThrow();
 
         if (project.getMetadata().getDeletionTimestamp() != null) {
             // project is being deleted
-            return notFound(notFoundStatusCode, "Tenant scheduled for deletion");
+            log.info("Tenant scheduled for deletion");
```
Comment (ctron) [inline iot/iot-device-registry-base/src/main/java/io/enmasse/iot/registry/tenant/KubernetesTenantInformationService.java:86]: It looks like the next call to `notFound` will again log the message.

File: iot/iot-device-registry-base/src/main/java/io/enmasse/iot/registry/tenant/KubernetesTenantInformationService.java [hunk 3]
```diff
@@ -73,7 +110,7 @@
      * @return The completed future, never returns {@code null}.
      */
     private static <T> Future<T> notFound(final int notFoundStatusCode, final String message) {
+        log.info(message);
```
Comment (ctron) [inline iot/iot-device-registry-base/src/main/java/io/enmasse/iot/registry/tenant/KubernetesTenantInformationService.java:113]: I don't think we should have "not found" in info, that might spam the logs a bit.

File: iot/iot-device-registry-base/src/main/java/io/enmasse/iot/registry/tenant/KubernetesTenantInformationService.java [hunk 4]
```diff
@@ -2,56 +2,93 @@
...
+                                                )
+                                        );
+                            } else {
+                                return Future.succeededFuture(Result.from(HTTP_NOT_FOUND));
+                            }
+                        }
+                );
+    }
+
+    public Future<TenantInformation> tenantExists(final String tenantName, final int notFoundStatusCode, final Span span) {
```
Comment (ctron) [inline iot/iot-device-registry-base/src/main/java/io/enmasse/iot/registry/tenant/KubernetesTenantInformationService.java:63]: It is hard for me to grasp the difference of the two `tenantExists` methods. I think Javadoc would help.

File: iot/iot-device-registry-base/src/main/java/io/enmasse/iot/registry/util/DeviceRegistryTokenAuthProvider.java [hunk 5]
```diff
@@ -62,7 +62,7 @@
     private final Cache<String, TokenReview> tokens;
     private final Cache<String, Boolean> authorizations;
 
-    protected TenantInformationService tenantInformationService;
+    protected KubernetesTenantInformationService tenantInformationService;
```
Comment (ctron) [inline iot/iot-device-registry-base/src/main/java/io/enmasse/iot/registry/util/DeviceRegistryTokenAuthProvider.java:65]: We might want to put in a note here to fix this upstream.

File: iot/iot-device-registry-infinispan/pom.xml [hunk 6]
```diff
@@ -113,6 +113,11 @@
         </exclusion>
       </exclusions>
     </dependency>
+      <dependency>
+          <groupId>org.eclipse.hono</groupId>
+          <artifactId>hono-service-device-registry-base</artifactId>
+          <version>1.2.1</version>
```
Comment (ctron) [inline iot/iot-device-registry-infinispan/pom.xml:119]: This should come from `hono.version` or a `dependencyManagement` entry.

File: iot/iot-infinispan-base/pom.xml [hunk 7]
```diff
@@ -78,6 +78,11 @@
       <groupId>org.hamcrest</groupId>
       <artifactId>hamcrest</artifactId>
     </dependency>
+      <dependency>
+          <groupId>org.eclipse.hono</groupId>
+          <artifactId>hono-service-device-registry-base</artifactId>
+          <version>1.2.1</version>
```
Comment (ctron) [inline iot/iot-infinispan-base/pom.xml:84]: Same here

File: iot/iot-jdbc-base/pom.xml [hunk 8]
```diff
@@ -76,6 +76,11 @@
       <artifactId>logback-classic</artifactId>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>org.eclipse.hono</groupId>
+      <artifactId>hono-service-device-registry-base</artifactId>
+      <version>1.2.1</version>
```
Comment (ctron) [inline iot/iot-jdbc-base/pom.xml:82]: Same here

File: templates/iot/examples/iot-config.yaml [hunk 9]
```diff
@@ -10,7 +10,7 @@ spec:
           external:
             username: registry
             password: user12
-            url: jdbc:postgresql://postgresql.device-registry-storage.svc/device-registry
```
Comment (ctron) [inline templates/iot/examples/iot-config.yaml:13]: JB fixed this in order to align this with the documentation. I think we should leave it that way.

File: templates/iot/examples/iot-config.yaml [hunk 10]
```diff
@@ -20,7 +20,7 @@ spec:
               connection:
                 username: registry
                 password: user12
-                url: jdbc:postgresql://postgresql.device-registry-storage.svc/device-registry
```
Comment (ctron) [inline templates/iot/examples/iot-config.yaml:23]: Same here

File: templates/iot/common/030-Deployment-iot-interior-interconnect.yaml [hunk 11]
```diff
@@ -0,0 +1,61 @@
+apiVersion: apps/v1
```
Comment (ctron) [inline templates/iot/common/030-Deployment-iot-interior-interconnect.yaml:1]: For all the deployments: I think it is ok to leave this for the moment, but we must extend the operator to take care of this.

File: iot/iot-device-registry-jdbc/src/main/java/io/enmasse/iot/registry/jdbc/config/DeviceConnectionServiceConfiguration.java [hunk 12]
```diff
@@ -18,7 +19,6 @@
 import io.vertx.core.Vertx;
 
 @Configuration
-@Profile(PROFILE_DEVICE_CONNECTION)
```
Comment (ctron) [inline iot/iot-device-registry-jdbc/src/main/java/io/enmasse/iot/registry/jdbc/config/DeviceConnectionServiceConfiguration.java:21]: I think dropping the profile here and the component annotation on the service, will break the deployment, as the operator can no longer properly fire up specific services.

File: iot/iot-device-registry-infinispan/src/main/java/io/enmasse/iot/registry/infinispan/device/impl/RegistrationServiceImpl.java [hunk 13]
```diff
@@ -62,6 +63,11 @@ public RegistrationServiceImpl(final DeviceManagementCacheProvider provider) {
 
     }
 
+    @Override
+    protected Future<JsonArray> resolveGroupMembers(String tenantId, JsonArray viaGroups, Span span) {
+        return null;
```
Comment (ctron) [inline iot/iot-device-registry-infinispan/src/main/java/io/enmasse/iot/registry/infinispan/device/impl/RegistrationServiceImpl.java:68]: We would at least add some `TODO`.

File: iot/iot-device-registry-infinispan/src/main/java/io/enmasse/iot/registry/infinispan/devcon/CacheDeviceConnectionService.java [hunk 14]
```diff
@@ -39,8 +39,6 @@
  * @see <a href="https://infinspan.org">https://infinspan.org</a>
  *
  */
-@Component
-@Profile(PROFILE_DEVICE_CONNECTION)
```
Comment (ctron) [inline iot/iot-device-registry-infinispan/src/main/java/io/enmasse/iot/registry/infinispan/devcon/CacheDeviceConnectionService.java:43]: Same here: the operator can no longer control the services being activated.

**Comment by dejanb** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.shared.control.CommandAndControlTest

**Comment by dejanb** [comment]
@enmasse-ci run tests profile=iot testcase=io.enmasse.systemtest.iot.shared.control.CommandAndControlTest

**Comment by dejanb** [comment]
@enmasse-ci run tests profile=iot

File: templates/iot/examples/iot-config.yaml (hunk 9 above)
Comment (dejanb) [reply templates/iot/examples/iot-config.yaml:13]: I would rather change the docs as how it is now, the getting started guide doesn't get you started :) (And you need figure on your own what's wrong)

**Comment by dejanb** [comment]
@enmasse-ci run tests profile=iot

**Comment by dejanb** [comment]
@enmasse-ci run tests profile=iot

**Comment by dejanb** [comment]
@enmasse-ci run tests ocp4 profile=iot

**Comment by dejanb** [comment]
@enmasse-ci run tests ocp4 profile=iot

File: iot/iot-device-registry-infinispan/pom.xml (hunk 6 above)
Comment (ctron) [reply iot/iot-device-registry-infinispan/pom.xml:119]: I think this still needs to be replaced.

File: iot/iot-device-registry-infinispan/src/main/java/io/enmasse/iot/registry/infinispan/config/DeviceConnectionServiceConfiguration.java [hunk 15]
```diff
@@ -29,4 +30,9 @@ public DeviceConnectionAmqpEndpoint deviceConnectionAmqpEndpoint(final Vertx ver
         return new DeviceConnectionAmqpEndpoint(vertx);
     }
 
+    @Bean
+    public MapBasedDeviceConnectionsConfigProperties deviceConnectionsProperties() {
+        return new MapBasedDeviceConnectionsConfigProperties();
```
Comment (ctron) [inline iot/iot-device-registry-infinispan/src/main/java/io/enmasse/iot/registry/infinispan/config/DeviceConnectionServiceConfiguration.java:35]: Add note for "fix me later"

**Comment by dejanb** [comment]
@enmasse-ci run tests ocp4 profile=iot

**Comment by dejanb** [comment]
@enmasse-ci run tests ocp4 profile=iot

**Comment by ctron** [comment]
I think we can merge this. The test failure is actually a timeout of the command deploying the SQL schema before the test. Which worked in all the different system tests though.

**Pull request merged**